src/errors.rs

```diff
@@ -1,50 +1,64 @@
-use super::Box;
+use super::{dealloc, Channel};
 use core::fmt;
 use core::mem;
+use core::ptr::NonNull;
 
 /// An error returned when trying to send on a closed channel. Returned from
 /// [`Sender::send`] if the corresponding [`Receiver`] has already been dropped.
 ///
 /// The message that could not be sent can be retreived again with [`SendError::into_inner`].
 pub struct SendError<T> {
-    channel_ptr: *mut super::Channel<T>,
+    channel_ptr: NonNull<Channel<T>>,
 }
 
 unsafe impl<T: Send> Send for SendError<T> {}
+unsafe impl<T: Sync> Sync for SendError<T> {}
 
 impl<T> SendError<T> {
-    pub(crate) const fn new(channel_ptr: *mut super::Channel<T>) -> Self {
+    /// Safety: by calling this function, the caller semantically transfers ownership of the
+    /// channel's resources to the created `SendError`. Thus the caller must ensure that the
+    /// pointer is not used in a way which would violate this ownership transfer. Moreover, the
+    /// channel must contain an initialized message.
+    pub(crate) const unsafe fn new(channel_ptr: NonNull<Channel<T>>) -> Self {
```

Comment:
I think we can consider making this `const unsafe fn new(channel: Box<Channel<T>>`. It would still be unsafe, since the caller must make sure the `message` is initialized, but that would remove a lot of the unsafety in this file regarding ownership and who frees the memory. The box going out of scope would automatically free the memory.

Reply:
I'm having second thoughts on this. In some future I want to investigate making this crate not have to depend on `alloc`, and allow it to work by having the user supply the storage for the `Channel`. With that in mind I don't want the error to have a stronger dependency on `Box` than it already has. Keep this as is!

Reply:
Will do!